exercises/concept/take-a-number-deluxe/lib/take_a_number_deluxe/state.ex

```diff
@@ -0,0 +1,85 @@
+defmodule TakeANumberDeluxe.State do
+  defstruct min_number: 1, max_number: 999, queue: nil, auto_shutdown_timeout: :infinity
+  @type t :: %__MODULE__{}
+
+  @spec new(integer, integer, timeout) :: {:ok, TakeANumberDeluxe.State.t()} | {:error, atom()}
+  def new(min_number, max_number, auto_shutdown_timeout \\ :infinity) do
+    if min_and_max_numbers_valid?(min_number, max_number) and
+         timeout_valid?(auto_shutdown_timeout) do
+      {:ok,
+       %__MODULE__{
+         min_number: min_number,
+         max_number: max_number,
+         queue: :queue.new(),
+         auto_shutdown_timeout: auto_shutdown_timeout
+       }}
+    else
+      {:error, :invalid_configuration}
+    end
+  end
+
+  @spec queue_new_number(TakeANumberDeluxe.State.t()) ::
+          {:ok, integer(), TakeANumberDeluxe.State.t()} | {:error, atom()}
+  def queue_new_number(%__MODULE__{} = state) do
+    case find_next_available_number(state) do
+      {:ok, next_available_number} ->
+        {:ok, next_available_number,
+         %{state | queue: :queue.in(next_available_number, state.queue)}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @spec serve_next_queued_number(TakeANumberDeluxe.State.t(), integer() | nil) ::
+          {:ok, integer(), TakeANumberDeluxe.State.t()} | {:error, atom()}
+  def serve_next_queued_number(%__MODULE__{} = state, priority_number) do
+    cond do
+      :queue.is_empty(state.queue) ->
+        {:error, :empty_queue}
+
+      is_nil(priority_number) ->
+        {{:value, next_number}, new_queue} = :queue.out(state.queue)
+        {:ok, next_number, %{state | queue: new_queue}}
+
+      :queue.member(priority_number, state.queue) ->
+        {:ok, priority_number, %{state | queue: :queue.delete(priority_number, state.queue)}}
```

Comment:
Looks like `:queue.delete` is only available from OTP 24. @jiegillet what would you propose as a fix? Back to my weird list-based queue? Or add a check with `function_exported?(:queue, :delete, 2)` and use `to_list`, `Enum.filter`, and `from_list` to fake `delete`?

Reply:
None of these suggestions sparked joy, so I implemented and pushed a queue module. Feel free to revert if it's not to your liking :)
I figured this exercise is already complicated, so this won't make a real difference :)

Reply:
Thank you! I'm glad you didn't make me implement a queue myself 😁. I just added an explanation comment at the top of the queue module.